docs/run_test/test_spec.rst

```diff
@@ -0,0 +1,184 @@
+Test Specification
```

Comment:
It looks can be auto generated. Remove it from the PR, and add the file to .gitignore, so it won't be committed suddenly.

Reply:
Will do after fixing bugs.

Reply:
```
Running Sphinx v4.1.2

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 327, in eval_config_file
    exec(code, namespace)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py", line 80, in <module>
    update_file(file_path)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/tools/test_spec_gen.py", line 26, in update_file
    with open(filename, "w") as test_spec:
NotADirectoryError: [Errno 20] Not a directory: '/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py/run_test/test_spec.rst'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/cmd/build.py", line 276, in build_main
    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/application.py", line 216, in __init__
    self.config = Config.read(self.confdir, confoverrides or {}, self.tags)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 172, in read
    namespace = eval_config_file(filename, tags)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 340, in eval_config_file
    raise ConfigError(msg % traceback.format_exc()) from exc
sphinx.errors.ConfigError: There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 327, in eval_config_file
    exec(code, namespace)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py", line 80, in <module>
    update_file(file_path)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/tools/test_spec_gen.py", line 26, in update_file
    with open(filename, "w") as test_spec:
NotADirectoryError: [Errno 20] Not a directory: '/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py/run_test/test_spec.rst'


Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 327, in eval_config_file
    exec(code, namespace)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py", line 80, in <module>
    update_file(file_path)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/tools/test_spec_gen.py", line 26, in update_file
    with open(filename, "w") as test_spec:
NotADirectoryError: [Errno 20] Not a directory: '/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py/run_test/test_spec.rst'
```Running Sphinx v4.1.2

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 327, in eval_config_file
    exec(code, namespace)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py", line 80, in <module>
    update_file(file_path)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/tools/test_spec_gen.py", line 26, in update_file
    with open(filename, "w") as test_spec:
NotADirectoryError: [Errno 20] Not a directory: '/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py/run_test/test_spec.rst'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/cmd/build.py", line 276, in build_main
    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/application.py", line 216, in __init__
    self.config = Config.read(self.confdir, confoverrides or {}, self.tags)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 172, in read
    namespace = eval_config_file(filename, tags)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 340, in eval_config_file
    raise ConfigError(msg % traceback.format_exc()) from exc
sphinx.errors.ConfigError: There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 327, in eval_config_file
    exec(code, namespace)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py", line 80, in <module>
    update_file(file_path)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/tools/test_spec_gen.py", line 26, in update_file
    with open(filename, "w") as test_spec:
NotADirectoryError: [Errno 20] Not a directory: '/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py/run_test/test_spec.rst'


Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/envs/wenyi-docs-tools/lib/python3.8/site-packages/sphinx/config.py", line 327, in eval_config_file
    exec(code, namespace)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py", line 80, in <module>
    update_file(file_path)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/tools/test_spec_gen.py", line 26, in update_file
    with open(filename, "w") as test_spec:
NotADirectoryError: [Errno 20] Not a directory: '/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/conf.py/run_test/test_spec.rst'

Reply:
It seems not to be able to create files when rtd is calling sphinx to build docs. Does that mean we need to keep the rST files in our repo or is there any feasible workaround?

Reply:
Found a solution using ``subprocess``. Source: [https://github.com/readthedocs/readthedocs.org/issues/2276](url)

Reply:
As the last line, may be it's a path issue?

NotADirectoryError: [Errno 20] Not a directory: '/home/docs/checkouts/readthedocs.org/user_builds/mslisa/checkouts/wenyi-docs-tools/docs/**conf.py**/run_test/test_spec.rst'

Reply:
I totally missed that. That makes sense.

Reply:
Btw when using `subprocess` the same issue happened as well, and it was fixed.